File: examples2d/platform2.rs
use rapier_testbed2d::Testbed;
use rapier2d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground.
     */
    let ground_size = 10.0;
    let ground_height = 0.1;

    let rigid_body = RigidBodyBuilder::fixed().translation(Vector::new(0.0, -ground_height));
    let collider = ColliderBuilder::cuboid(ground_size, ground_height);
    let _ = world.insert(rigid_body, collider);

    /*
     * Create the boxes
     */
    let num = 6;
    let rad = 0.2;

    let shift = rad * 2.0;
    let centerx = shift * num as f32 / 2.0;
    let centery = shift / 2.0 + 3.04;

    for i in 0usize..num {
        for j in 0usize..num {
            let x = i as f32 * shift - centerx;
            let y = j as f32 * shift + centery;

            // Build the rigid body.
            let rigid_body = RigidBodyBuilder::dynamic().translation(Vector::new(x, y));
            let collider = ColliderBuilder::cuboid(rad, rad);
            let _ = world.insert(rigid_body, collider);
        }
    }

    /*
     * Setup a position-based kinematic rigid body.
     */
    let platform_body = RigidBodyBuilder::kinematic_velocity_based()
        .translation(Vector::new(-10.0 * rad, 1.5 + 0.8));
    let collider = ColliderBuilder::cuboid(rad * 10.0, rad);
    let (velocity_based_platform_handle, _) = world.insert(platform_body, collider);

    /*
     * Setup a velocity-based kinematic rigid body.
     */
    let platform_body = RigidBodyBuilder::kinematic_position_based()
        .translation(Vector::new(-10.0 * rad, 2.0 + 1.5 + 0.8));
    let collider = ColliderBuilder::cuboid(rad * 10.0, rad);
    let (position_based_platform_handle, _) = world.insert(platform_body, collider);

    /*
     * Setup a callback to control the platform.
     */
    testbed.add_callback(move |_, physics, _, run_state| {
        let velocity = Vector::new(run_state.time.sin() * 5.0, (run_state.time * 5.0).sin());

        // Update the velocity-based kinematic body by setting its velocity.
        if let Some(platform) = physics.bodies.get_mut(velocity_based_platform_handle) {
            platform.set_linvel(velocity, true);
        }

        // Update the position-based kinematic body by setting its next position.
        if let Some(platform) = physics.bodies.get_mut(position_based_platform_handle) {
            let mut next_tra = platform.translation();
            next_tra += velocity * physics.integration_parameters.dt;
            platform.set_next_kinematic_translation(next_tra);
        }
    });

    /*
     * Run the simulation.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec2::new(0.0, 1.0), 40.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/platform2.py
"""Port of examples2d/platform2.rs."""
from __future__ import annotations

import math

import rapier2d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Platform"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 10.0
    ground_height = 0.1

    handle = bodies.insert(rp.RigidBody.fixed().translation((0.0, -ground_height)))
    colliders.insert_with_parent(
        rp.Collider.cuboid(ground_size, ground_height), handle, bodies
    )

    num = 6
    rad = 0.2
    shift = rad * 2.0
    centerx = shift * num / 2.0
    centery = shift / 2.0 + 3.04

    for i in range(num):
        for j in range(num):
            x = i * shift - centerx
            y = j * shift + centery
            handle = bodies.insert(rp.RigidBody.dynamic().translation((x, y)))
            colliders.insert_with_parent(rp.Collider.cuboid(rad, rad), handle, bodies)

    velocity_based_platform_handle = bodies.insert(
        rp.RigidBody.kinematic_velocity_based().translation((-10.0 * rad, 1.5 + 0.8))
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(rad * 10.0, rad), velocity_based_platform_handle, bodies
    )

    position_based_platform_handle = bodies.insert(
        rp.RigidBody.kinematic_position_based().translation(
            (-10.0 * rad, 2.0 + 1.5 + 0.8)
        )
    )
    colliders.insert_with_parent(
        rp.Collider.cuboid(rad * 10.0, rad), position_based_platform_handle, bodies
    )

    state = {"t": 0.0}

    def callback(tb) -> None:
        dt = tb._integration_parameters.dt
        state["t"] += dt
        velocity = (math.sin(state["t"]) * 5.0, math.sin(state["t"] * 5.0))

        platform = tb.bodies.get(velocity_based_platform_handle)
        if platform is not None:
            platform.linvel = velocity

        platform = tb.bodies.get(position_based_platform_handle)
        if platform is not None:
            tr = platform.translation
            new_pos = rp.Isometry2(
                (float(tr.x) + velocity[0] * dt, float(tr.y) + velocity[1] * dt),
                0.0,
            )
            platform.position = new_pos

    testbed.add_callback(callback)
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 1.0), zoom=40.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
